rosbag2_transport/src/rosbag2_transport/rosbag2_transport_python.cpp

```diff
@@ -0,0 +1,134 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Python.h>
+#include <string>
+#include <vector>
+
+#include "rosbag2_transport/rosbag2_transport.hpp"
+#include "rosbag2_transport/record_options.hpp"
+#include "rosbag2_transport/storage_options.hpp"
+
+static PyObject *
+rosbag2_transport_record(PyObject * Py_UNUSED(self), PyObject * args, PyObject * kwargs)
+{
+  rosbag2_transport::StorageOptions storage_options{};
+  rosbag2_transport::RecordOptions record_options{};
+
+  static const char * kwlist[] = {"uri", "storage_id", "all", "topics", nullptr};
+
+  char * uri = nullptr;
+  char * storage_id = nullptr;
+  bool all = false;
+  PyObject * topics = nullptr;
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "ss|bO", const_cast<char **>(kwlist),
+    &uri,
+    &storage_id,
+    &all,
+    &topics))
+  {
+    return nullptr;
+  }
+
+  storage_options.uri = std::string(uri);
+  storage_options.storage_id = std::string(storage_id);
+  record_options.all = all;
+
+  if (topics) {
+    PyObject * topic_iterator = PyObject_GetIter(topics);
+    if (topic_iterator != nullptr) {
+      PyObject * topic;
+      while ((topic = PyIter_Next(topic_iterator))) {
+        record_options.topics.emplace_back(PyUnicode_AsUTF8(topic));
+
+        Py_DECREF(topic);
+      }
+      Py_DECREF(topic_iterator);
+    }
+  }
+
+  rosbag2_transport::Rosbag2Transport transport;
+  transport.init();
+  transport.record(storage_options, record_options);
+  transport.shutdown();
+
+  Py_RETURN_NONE;
+}
+
+static PyObject *
+rosbag2_transport_play(PyObject * Py_UNUSED(self), PyObject * args, PyObject * kwargs)
+{
+  rosbag2_transport::PlayOptions play_options{};
+  rosbag2_transport::StorageOptions storage_options{};
+
+  static const char * kwlist[] = {"uri", "storage_id", "read_ahead_queue_size", nullptr};
+
+  char * uri;
+  char * storage_id;
+  size_t read_ahead_queue_size;
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "ss|k", const_cast<char **>(kwlist),
```

Comment:
Does this function return `false` when any of the keywords are not defined? Because if so, then I believe the check below of
```
play_options.read_ahead_queue_size = read_ahead_queue_size ? read_ahead_queue_size : 1000;
```
does not have any effect.
In the case this function does return `true` even if some keywords are not available, then I think we have to check the value for `uri` and `storage_id` as well.

Reply:
Only for non-optional keywords. Anything behind the `|` is optional. So the check below is necessary. It will return `false` if the mandatory arguments `uri` and `storage_id` are missing.